storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java

```diff
@@ -126,4 +128,56 @@ private boolean isFinalized(final UnsignedLong slot) {
   public Optional<BeaconState> getNonfinalizedBlockState(final Bytes32 blockRoot) {
     return recentChainData.getBlockState(blockRoot);
   }
+
+  /**
+   * Returns the state which was proposed in or most recently before the requested slot on the chain
+   * specified by <code>headBlockRoot</code>. If the slot was empty, the state at the last filled
+   * slot is returned.
```

Comment:
You can't propose a state - every slot just has a state.  We should check what other clients do when requesting the state for a slot that doesn't contain a block.  Do they return the state from the last full slot or do they actually process the skipped slots?  The block_root is the same in both cases (because there's no new block) but the state root and state slot number would be different.

Reply:
copy / paste :/ i'll clean that up